python/lsst/jointcal/jointcal.py

```diff
@@ -419,32 +503,260 @@ class JointcalTask(pipeBase.PipelineTask, pipeBase.CmdLineTask):
 
-    def __init__(self, butler=None, profile_jointcal=False, **kwargs):
+    def __init__(self, butler=None, initInputs=None, **kwargs):
         super().__init__(**kwargs)
-        self.profile_jointcal = profile_jointcal
         self.makeSubtask("sourceSelector")
         if self.config.doAstrometry:
-            self.makeSubtask('astrometryRefObjLoader', butler=butler)
+            if initInputs is None:
+                # gen3 middleware does refcat things internally (and will not have a butler here)
+                self.makeSubtask('astrometryRefObjLoader', butler=butler)
             self.makeSubtask("astrometryReferenceSelector")
         else:
             self.astrometryRefObjLoader = None
         if self.config.doPhotometry:
-            self.makeSubtask('photometryRefObjLoader', butler=butler)
+            if initInputs is None:
+                # gen3 middleware does refcat things internally (and will not have a butler here)
+                self.makeSubtask('photometryRefObjLoader', butler=butler)
             self.makeSubtask("photometryReferenceSelector")
         else:
             self.photometryRefObjLoader = None
 
         # To hold various computed metrics for use by tests
         self.job = Job.load_metrics_package(subset='jointcal')
 
+    def runQuantum(self, butlerQC, inputRefs, outputRefs):
+        # We override runQuantum to set up the refObjLoaders and write the
+        # outputs to the correct refs.
+        inputs = butlerQC.get(inputRefs)
+        # We want the tract number for writing debug files
+        tract = butlerQC.quantum.dataId['tract']
+        if self.config.doAstrometry:
+            self.astrometryRefObjLoader = ReferenceObjectLoader(
+                dataIds=[ref.datasetRef.dataId
+                         for ref in inputRefs.astrometryRefCat],
+                refCats=inputs.pop('astrometryRefCat'),
+                config=self.config.astrometryRefObjLoader,
+                log=self.log)
+        if self.config.doPhotometry:
+            self.photometryRefObjLoader = ReferenceObjectLoader(
+                dataIds=[ref.datasetRef.dataId
+                         for ref in inputRefs.photometryRefCat],
+                refCats=inputs.pop('photometryRefCat'),
+                config=self.config.photometryRefObjLoader,
+                log=self.log)
+        outputs = self.run(**inputs, tract=tract)
+        if self.config.doAstrometry:
+            self._put_output(butlerQC, outputs.outputWcs, outputRefs.outputWcs)
+        if self.config.doPhotometry:
+            self._put_output(butlerQC, outputs.outputPhotoCalib, outputRefs.outputPhotoCalib)
+
+    def _put_output(self, butlerQC, outputs, outputRefs):
+        """Butler put the output datasets to their appropriate datarefs.
+
+        Parameters
+        ----------
+        butlerQC : `ButlerQuantumContext`
+            A butler which is specialized to operate in the context of a
+            `lsst.daf.butler.Quantum`.
```

Comment:
Is this different than the standard Gen 3 Butler?

Reply:
Yes, though I don't really know how. I'm just copying from the definition in the docstring for pipeBase's `PipelineTask.runQuantum()`. I've added a note.